resourcettl/cache.go

```diff
@@ -0,0 +1,187 @@
+	checkouts      map[string]int      // keeps track of how many checkouts are there for a given instanceID
+	expiries       map[string]struct{} // keeps track of instances that are expired and need to be cleaned up after all checkouts are done
+
+	ttl      time.Duration
+	ttlcache *cachettl.Cache[K, string]
+}
+
+// Checkout returns a resource for the given key. If the resource is not available, it creates a new one.
+// The caller must call the returned checkin function when the resource is no longer needed, to release the resource.
+// Multiple checkouts for the same key are allowed and they can all share the same resource. The resource is cleaned up
+// only when all checkouts are checked-in and the resource's ttl has expired (or its key has been invalidated through [Invalidate]).
+func (c *Cache[K, R]) Checkout(key K) (resource R, checkin func(), err error) {
+	defer c.lockKey(key)()
+
+	if instanceID := c.ttlcache.Get(key); instanceID != "" {
+		c.mu.Lock()
+		defer c.mu.Unlock()
+		r := c.resources[instanceID]
+		c.checkouts[instanceID]++
+		return r, c.checkinFunc(key, r, instanceID), nil
+	}
+	return c.newInstance(key)
+}
+
+// Invalidate invalidates the resource for the given key.
+func (c *Cache[K, R]) Invalidate(key K) {
+	defer c.lockKey(key)()
+	c.mu.Lock()
+	instanceID, ok := c.latestInstance[key]
+	c.mu.Unlock()
+	if ok {
+		c.ttlcache.Put(key, "", 0)
+		c.onEvicted(key, instanceID)
+	}
+}
+
+// newInstance creates a new resource for the given key.
+func (c *Cache[K, R]) newInstance(key K) (R, func(), error) {
+	r, err := c.new(key)
+	if err != nil {
+		return r, nil, err
+	}
+
+	instanceID := uuid.NewString()
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.latestInstance[key] = instanceID
+	c.resources[instanceID] = r
+	c.checkouts[instanceID]++
+	c.ttlcache.Put(key, instanceID, c.ttl)
+	return r, c.checkinFunc(key, r, instanceID), nil
+}
+
+// checkinFunc returns a function that decrements the checkout count and cleans up the resource if it is no longer needed.
+func (c *Cache[K, R]) checkinFunc(key K, r R, instanceID string) func() {
+	var once sync.Once
+	return func() {
+		once.Do(func() {
+			c.mu.Lock()
+			defer c.mu.Unlock()
+			c.checkouts[instanceID]--
+			if _, ok := c.expiries[instanceID]; ok && // instance is expired
+				c.checkouts[instanceID] == 0 { // no more checkouts
+				delete(c.expiries, instanceID)
+				delete(c.latestInstance, key)
+				// not deleting the latestInstance entry, as it might be used by another instance
+				go c.cleanup(r)
+			}
+		})
+	}
+}
+
+// onEvicted is called when a key is evicted from the cache. It cleans up the resource if it is not checked out.
+func (c *Cache[K, R]) onEvicted(_ K, instanceID string) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	checkouts, ok := c.checkouts[instanceID]
+	if !ok {
+		return // already cleaned up through invalidate
+	}
+	if checkouts == 0 {
+		r := c.resources[instanceID]
+		delete(c.resources, instanceID)
+		delete(c.checkouts, instanceID)
+		go c.cleanup(r)
```

Comment:
Could the lack of synchronization create issues here? Wdyt?

Reply:
Using another go-routine so that the main go-routine is not blocked. At this point no one should be using the resource, thus doing async cleanup should be safe